release/src/router/openvpn/doc/openvpn.8
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "OPENVPN" 8 "" "" "System Manager's Manual"
.SH NAME
openvpn \- Secure IP tunnel daemon
.SH SYNOPSIS
.nf
\fBopenvpn\fP [ options ... ]
\fBopenvpn\fP  \fB\-\-help\fP
.fi
.sp
.SH INTRODUCTION
.sp
OpenVPN is an open source VPN daemon by James Yonan. Because OpenVPN
tries to be a universal VPN tool offering a great deal of flexibility,
there are a lot of options on this manual page. If you\(aqre new to
OpenVPN, you might want to skip ahead to the examples section where you
will see how to construct simple VPNs on the command line without even
needing a configuration file.
.sp
Also note that there\(aqs more documentation and examples on the OpenVPN
web site: \fI\%https://openvpn.net/\fP
.sp
And if you would like to see a shorter version of this manual, see the
openvpn usage message which can be obtained by running \fBopenvpn\fP
without any parameters.
.SH DESCRIPTION
.sp
OpenVPN is a robust and highly flexible VPN daemon. OpenVPN supports
SSL/TLS security, ethernet bridging, TCP or UDP tunnel transport through
proxies or NAT, support for dynamic IP addresses and DHCP, scalability
to hundreds or thousands of users, and portability to most major OS
platforms.
.sp
OpenVPN is tightly bound to the OpenSSL library, and derives much of its
crypto capabilities from it.
.sp
OpenVPN supports conventional encryption using a pre\-shared secret key
\fB(Static Key mode)\fP or public key security \fB(SSL/TLS mode)\fP using
client & server certificates. OpenVPN also supports non\-encrypted
TCP/UDP tunnels.
.sp
OpenVPN is designed to work with the \fBTUN/TAP\fP virtual networking
interface that exists on most platforms.
.sp
Overall, OpenVPN aims to offer many of the key features of IPSec but
with a relatively lightweight footprint.
.SH OPTIONS
.sp
OpenVPN allows any option to be placed either on the command line or in
a configuration file. Though all command line options are preceded by a
double\-leading\-dash (\(dq\-\-\(dq), this prefix can be removed when an option is
placed in a configuration file.
.SS Generic Options
.sp
This section covers generic options which are accessible regardless of
which mode OpenVPN is configured as.
.INDENT 0.0
.TP
.B  \-\-help
Show options.
.TP
.B  \-\-auth\-nocache
Don\(aqt cache \fB\-\-askpass\fP or \fB\-\-auth\-user\-pass\fP username/passwords in
virtual memory.
.sp
If specified, this directive will cause OpenVPN to immediately forget
username/password inputs after they are used. As a result, when OpenVPN
needs a username/password, it will prompt for input from stdin, which
may be multiple times during the duration of an OpenVPN session.
.sp
When using \fB\-\-auth\-nocache\fP in combination with a user/password file
and \fB\-\-chroot\fP or \fB\-\-daemon\fP, make sure to use an absolute path.
.TP
.BI \-\-cd \ dir
Change directory to \fBdir\fP prior to reading any files such as
configuration files, key files, scripts, etc. \fBdir\fP should be an
absolute path, with a leading \(dq/\(dq, and without any references to the
current directory such as \fB\&.\fP or \fB\&..\fP\&.
.sp
This option is useful when you are running OpenVPN in \fB\-\-daemon\fP mode,
and you want to consolidate all of your OpenVPN control files in one
location.
.TP
.BI \-\-chroot \ dir
Chroot to \fBdir\fP after initialization. \fB\-\-chroot\fP essentially
redefines \fBdir\fP as being the top level directory tree (/). OpenVPN
will therefore be unable to access any files outside this tree. This can
be desirable from a security standpoint.
.sp
Since the chroot operation is delayed until after initialization, most
OpenVPN options that reference files will operate in a pre\-chroot
context.
.sp
In many cases, the \fBdir\fP parameter can point to an empty directory,
however complications can result when scripts or restarts are executed
after the chroot operation.
.sp
Note: The SSL library will probably need \fB/dev/urandom\fP to be available
inside the chroot directory \fBdir\fP\&. This is because SSL libraries
occasionally need to collect fresh randomness. Newer linux kernels and some
BSDs implement a getrandom() or getentropy() syscall that removes the
need for \fB/dev/urandom\fP to be available.
.TP
.BI \-\-compat\-mode \ version
This option provides a convenient way to alter the defaults of OpenVPN
to be more compatible with the version \fBversion\fP specified. All of
the changes this option applies can also be achieved using individual
configuration options.
.sp
The version specified with this option is the version of OpenVPN peer
OpenVPN should try to be compatible with. In general OpenVPN should be
compatible with the last two previous version without this option. E.g.
OpenVPN 2.6.0 should be compatible with 2.5.x and 2.4.x without this option.
However, there might be some edge cases that still require this option even
in these cases.
.sp
Note: Using this option reverts defaults to no longer recommended
values and should be avoided if possible.
.sp
The following table details what defaults are changed depending on the
version specified.
.INDENT 7.0
.IP \(bu 2
2.5.x or lower: \fB\-\-allow\-compression asym\fP is automatically added
to the configuration if no other compression options are present.
.IP \(bu 2
2.4.x or lower: The cipher in \fB\-\-cipher\fP is appended to
\fB\-\-data\-ciphers\fP\&.
.IP \(bu 2
2.3.x or lower: \fB\-\-data\-ciphers\-fallback\fP is automatically added with
the same cipher as \fB\-\-cipher\fP\&.
.IP \(bu 2
2.3.6 or lower: \fB\-\-tls\-version\-min 1.0\fP is added to the configuration
when \fB\-\-tls\-version\-min\fP is not explicitly set.
.UNINDENT
.sp
If not required, this is option should be avoided. Setting this option can
lower security or disable features like data\-channel offloading.
.TP
.BI \-\-config \ file
Load additional config options from \fBfile\fP where each line corresponds
to one command line option, but with the leading \fB\-\-\fP removed.
.sp
If \fB\-\-config file\fP is the only option to the openvpn command, the
\fB\-\-config\fP can be removed, and the command can be given as \fBopenvpn
file\fP
.sp
Note that configuration files can be nested to a reasonable depth.
.sp
Double quotation or single quotation characters (\(dq\(dq, \(aq\(aq) can be used to
enclose single parameters containing whitespace, and \(dq#\(dq or \(dq;\(dq
characters in the first column can be used to denote comments.
.sp
Note that OpenVPN 2.0 and higher performs backslash\-based shell escaping
for characters not in single quotations, so the following mappings
should be observed:
.INDENT 7.0
.INDENT 3.5
.sp
.EX
\e\e       Maps to a single backslash character (\e).
\e\(dq       Pass a literal doublequote character (\(dq), don\(aqt
         interpret it as enclosing a parameter.
\e[SPACE] Pass a literal space or tab character, don\(aqt
         interpret it as a parameter delimiter.
.EE
.UNINDENT
.UNINDENT
.sp
For example on Windows, use double backslashes to represent pathnames:
.INDENT 7.0
.INDENT 3.5
.sp
.EX
secret \(dqc:\e\eOpenVPN\e\esecret.key\(dq
.EE
.UNINDENT
.UNINDENT
.sp
For examples of configuration files, see
\fI\%https://openvpn.net/community\-resources/how\-to/\fP
.sp
Here is an example configuration file:
.INDENT 7.0
.INDENT 3.5
.sp
.EX
#
# Sample OpenVPN configuration file for
# using a pre\-shared static key.
#
# \(aq#\(aq or \(aq;\(aq may be used to delimit comments.

# Use a dynamic tun device.
dev tun

# Our remote peer
remote mypeer.mydomain

# 10.1.0.1 is our local VPN endpoint
# 10.1.0.2 is our remote VPN endpoint
ifconfig 10.1.0.1 10.1.0.2

# Our pre\-shared static key
secret static.key
.EE
.UNINDENT
.UNINDENT
.TP
.BI \-\-daemon \ progname
Become a daemon after all initialization functions are completed.
.sp
Valid syntaxes:
.INDENT 7.0
.INDENT 3.5
.sp
.EX
daemon
daemon progname
.EE
.UNINDENT
.UNINDENT
.sp
This option will cause all message and error output to be sent to the syslog
file (such as \fB/var/log/messages\fP), except for the output of
scripts and ifconfig commands, which will go to \fB/dev/null\fP unless
otherwise redirected. The syslog redirection occurs immediately at the
point that \fB\-\-daemon\fP is parsed on the command line even though the
daemonization point occurs later. If one of the \fB\-\-log\fP options is
present, it will supersede syslog redirection.
.sp
The optional \fBprogname\fP parameter will cause OpenVPN to report its
program name to the system logger as \fBprogname\fP\&. This can be useful in
linking OpenVPN messages in the syslog file with specific tunnels. When
unspecified, \fBprogname\fP defaults to \fBopenvpn\fP\&.
.sp
When OpenVPN is run with the \fB\-\-daemon\fP option, it will try to delay
daemonization until the majority of initialization functions which are
capable of generating fatal errors are complete. This means that
initialization scripts can test the return status of the openvpn command
for a fairly reliable indication of whether the command has correctly
initialized and entered the packet forwarding event loop.
.sp
In OpenVPN, the vast majority of errors which occur after initialization
are non\-fatal.
.sp
Note: as soon as OpenVPN has daemonized, it can not ask for usernames,
passwords, or key pass phrases anymore. This has certain consequences,
namely that using a password\-protected private key will fail unless the
\fB\-\-askpass\fP option is used to tell OpenVPN to ask for the pass phrase.
.sp
Further, using \fB\-\-daemon\fP together with \fB\-\-auth\-user\-pass\fP (entered
on console) and \fB\-\-auth\-nocache\fP will fail as soon as key
renegotiation (and reauthentication) occurs.
.TP
.B  \-\-disable\-occ
\fBDEPRECATED\fP Disable \(dqoptions consistency check\(dq (OCC) in configurations
that do not use TLS.
.sp
Don\(aqt output a warning message if option inconsistencies are detected
between peers. An example of an option inconsistency would be where one
peer uses \fB\-\-dev tun\fP while the other peer uses \fB\-\-dev tap\fP\&.
.sp
Use of this option is discouraged, but is provided as a temporary fix in
situations where a recent version of OpenVPN must connect to an old
version.
.TP
.BI \-\-engine \ engine\-name
Enable OpenSSL hardware\-based crypto engine functionality.
.sp
Valid syntaxes:
.INDENT 7.0
.INDENT 3.5
.sp
.EX
engine
engine engine\-name
.EE
.UNINDENT
.UNINDENT
.sp
If \fBengine\-name\fP is specified, use a specific crypto engine. Use the
\fB\-\-show\-engines\fP standalone option to list the crypto engines which
are supported by OpenSSL.
.TP
.BI \-\-group \ group
Similar to the \fB\-\-user\fP option, this option changes the group ID of
the OpenVPN process to \fBgroup\fP after initialization.
.TP
.BI \-\-ignore\-unknown\-option \ args
Valid syntax:
.INDENT 7.0
.INDENT 3.5
.sp
.EX
ignore\-unknown\-option opt1 opt2 opt3 ... optN
.EE
.UNINDENT
.UNINDENT
.sp
When one of options \fBopt1 ... optN\fP is encountered in the configuration
file the configuration file parsing does not fail if this OpenVPN version
does not support the option. Multiple \fB\-\-ignore\-unknown\-option\fP options
can be given to support a larger number of options to ignore.
.sp
This option should be used with caution, as there are good security
reasons for having OpenVPN fail if it detects problems in a config file.
Having said that, there are valid reasons for wanting new software
features to gracefully degrade when encountered by older software
versions.
.sp
\fB\-\-ignore\-unknown\-option\fP is available since OpenVPN 2.3.3.
.TP
.BI \-\-iproute \ cmd
Set alternate command to execute instead of default \fBiproute2\fP command.
May be used in order to execute OpenVPN in unprivileged environment.
.TP
.BI \-\-keying\-material\-exporter \ args
Save Exported Keying Material [RFC5705] of \fBlen\fP bytes (must be between 16
and 4095 bytes) using \fBlabel\fP in environment
(\fBexported_keying_material\fP) for use by plugins in
\fBOPENVPN_PLUGIN_TLS_FINAL\fP callback.
.sp
Valid syntax:
.INDENT 7.0
.INDENT 3.5
.sp
.EX
keying\-material\-exporter label len
.EE
.UNINDENT
.UNINDENT
.sp
Note that exporter \fBlabels\fP have the potential to collide with existing
PRF labels. In order to prevent this, labels \fIMUST\fP begin with
\fBEXPORTER\fP\&.
.TP
.B  \-\-mlock
Disable paging by calling the POSIX mlockall function. Requires that
OpenVPN be initially run as root (though OpenVPN can subsequently
downgrade its UID using the \fB\-\-user\fP option).
.sp
Using this option ensures that key material and tunnel data are never
written to disk due to virtual memory paging operations which occur
under most modern operating systems. It ensures that even if an attacker
was able to crack the box running OpenVPN, he would not be able to scan
the system swap file to recover previously used ephemeral keys, which
are used for a period of time governed by the \fB\-\-reneg\fP options (see
below), then are discarded.
.sp
The downside of using \fB\-\-mlock\fP is that it will reduce the amount of
physical memory available to other applications.
.sp
The limit on how much memory can be locked and how that limit
is enforced are OS\-dependent. On Linux the default limit that an
unprivileged process may lock (RLIMIT_MEMLOCK) is low, and if
privileges are dropped later, future memory allocations will very
likely fail. The limit can be increased using ulimit or systemd
directives depending on how OpenVPN is started.
.sp
If the platform has the getrlimit(2) system call, OpenVPN will check
for the amount of mlock\-able memory before calling mlockall(2), and
tries to increase the limit to 100 MB if less than this is configured.
100 Mb is somewhat arbitrary \- it is enough for a moderately\-sized
OpenVPN deployment, but the memory usage might go beyond that if the
number of concurrent clients is high.
.TP
.BI \-\-nice \ n
Change process priority after initialization (\fBn\fP greater than 0 is
lower priority, \fBn\fP less than zero is higher priority).
.TP
.BI \-\-providers \ providers
Load the list of (OpenSSL) providers. This is mainly useful for using an
external provider for key management like tpm2\-openssl or to load the
legacy provider with
.INDENT 7.0
.INDENT 3.5
.sp
.EX
\-\-providers legacy default
.EE
.UNINDENT
.UNINDENT
.sp
Behaviour of changing this option between \fBSIGHUP\fP might not be well behaving.
If you need to change/add/remove this option, fully restart OpenVPN.
.TP
.BI \-\-remap\-usr1 \ signal
Control whether internally or externally generated \fBSIGUSR1\fP signals
are remapped to \fBSIGHUP\fP (restart without persisting state) or
\fBSIGTERM\fP (exit).
.sp
\fBsignal\fP can be set to \fBSIGHUP\fP or \fBSIGTERM\fP\&. By default,
no remapping occurs.
.TP
.BI \-\-script\-security \ level
This directive offers policy\-level control over OpenVPN\(aqs usage of
external programs and scripts. Lower \fBlevel\fP values are more
restrictive, higher values are more permissive. Settings for \fBlevel\fP:
.INDENT 7.0
.TP
.B \fB0\fP
Strictly no calling of external programs.
.TP
.B \fB1\fP
(Default) Only call built\-in executables such as ifconfig,
ip, route, or netsh.
.TP
.B \fB2\fP
Allow calling of built\-in executables and user\-defined
scripts.
.TP
.B \fB3\fP
Allow passwords to be passed to scripts via environmental
variables (potentially unsafe).
.UNINDENT
.sp
OpenVPN releases before v2.3 also supported a \fBmethod\fP flag which
indicated how OpenVPN should call external commands and scripts. This
could be either \fBexecve\fP or \fBsystem\fP\&. As of OpenVPN 2.3, this
flag is no longer accepted.
.sp
Some directives such as \fB\-\-up\fP allow options to be passed to the
external script. In these cases make sure the script name does not
contain any spaces or the configuration parser will choke because it
can\(aqt determine where the script name ends and script options start.
.sp
On Windoes it is a strict requirement to have the full path to the script
interpreter when running non\-executables files. This is not needed for
executable files, such as .exe, .com, .bat or .cmd files. For example,
if you have a Visual Basic script, you must use this syntax:
.INDENT 7.0
.INDENT 3.5
.sp
.EX
\-\-up \(aqC:\e\eWindows\e\eSystem32\e\ewscript.exe C:\e\eProgram\e Files\e\eOpenVPN\e\econfig\e\emy\-up\-script.vbs\(aq
.EE
.UNINDENT
.UNINDENT
.sp
Please note the single quote marks and the escaping of the backslashes
(\fB\e\e\fP) and the space character.
.TP
.BI \-\-setcon \ context
Apply SELinux \fBcontext\fP after initialization. This essentially
provides the ability to restrict OpenVPN\(aqs rights to only network I/O
operations, thanks to SELinux. This goes further than \fB\-\-user\fP and
\fB\-\-chroot\fP in that those two, while being great security features,
unfortunately do not protect against privilege escalation by
exploitation of a vulnerable system call. You can of course combine all
three, but please note that since setcon requires access to /proc you
will have to provide it inside the chroot directory (e.g. with mount
\-\-bind).
.sp
Since the setcon operation is delayed until after initialization,
OpenVPN can be restricted to just network\-related system calls, whereas
by applying the context before startup (such as the OpenVPN one provided
in the SELinux Reference Policies) you will have to allow many things
required only during initialization.
.sp
Like with chroot, complications can result when scripts or restarts are
executed after the setcon operation, which is why you should really
consider using the \fB\-\-persist\-tun\fP option.
.TP
.BI \-\-status \ args
Write operational status to \fBfile\fP every \fBn\fP seconds. \fBn\fP defaults
to \fB60\fP if not specified.
.sp
Valid syntaxes:
.INDENT 7.0
.INDENT 3.5
.sp
.EX
status file
status file n
.EE
.UNINDENT
.UNINDENT
.sp
Status can also be written to the syslog by sending a \fBSIGUSR2\fP
signal.
.sp
With multi\-client capability enabled on a server, the status file
includes a list of clients and a routing table. The output format can be
controlled by the \fB\-\-status\-version\fP option in that case.
.sp
For clients or instances running in point\-to\-point mode, it will contain
the traffic statistics.
.TP
.BI \-\-status\-version \ n
Set the status file format version number to \fBn\fP\&.
.sp
This only affects the status file on servers with multi\-client
capability enabled.  Valid status version values:
.INDENT 7.0
.TP
.B \fB1\fP
Traditional format (default). The client list contains the
following fields comma\-separated: Common Name, Real Address, Bytes
Received, Bytes Sent, Connected Since.
.TP
.B \fB2\fP
A more reliable format for external processing. Compared to
version \fB1\fP, the client list contains some additional fields:
Virtual Address, Virtual IPv6 Address, Username, Client ID, Peer ID,
Data Channel Cipher. Future versions may extend the number of fields.
.TP
.B \fB3\fP
Identical to \fB2\fP, but fields are tab\-separated.
.UNINDENT
.TP
.B  \-\-test\-crypto
Do a self\-test of OpenVPN\(aqs crypto options by encrypting and decrypting
test packets using the data channel encryption options specified above.
This option does not require a peer to function, and therefore can be
specified without \fB\-\-dev\fP or \fB\-\-remote\fP\&.
.sp
The typical usage of \fB\-\-test\-crypto\fP would be something like this:
.INDENT 7.0
.INDENT 3.5
.sp
.EX
openvpn \-\-test\-crypto
.EE
.UNINDENT
.UNINDENT
.sp
or
.INDENT 7.0
.INDENT 3.5
.sp
.EX
openvpn \-\-test\-crypto \-\-verb 9
.EE
.UNINDENT
.UNINDENT
.sp
This option is very useful to test OpenVPN after it has been ported to a
new platform, or to isolate problems in the compiler, OpenSSL crypto
library, or OpenVPN\(aqs crypto code. Since it is a self\-test mode,
problems with encryption and authentication can be debugged
independently of network and tunnel issues.
.sp
Older versions of OpenVPN used the \fB\-\-secret\fP argument to specify a
static key for this test. Newer version generate a random key for the
test.
.TP
.BI \-\-tmp\-dir \ dir
Specify a directory \fBdir\fP for temporary files instead of the default
\fBTMPDIR\fP (or \(dq/tmp\(dq if unset). Note that it must be writable by the main
process after it has dropped root privileges.
.sp
This directory will be used to communicate with scripts and plugins:
.INDENT 7.0
.IP \(bu 2
\fB\-\-client\-connect\fP scripts and \fBOPENVPN_PLUGIN_CLIENT_CONNECT\fP
plug\-in hook to dynamically generate client\-specific configuration
\fBclient_connect_config_file\fP and return success/failure via
\fBclient_connect_deferred_file\fP when using deferred client connect
method
.IP \(bu 2
\fBOPENVPN_PLUGIN_AUTH_USER_PASS_VERIFY\fP plug\-in hooks returns
success/failure via \fBauth_control_file\fP when using deferred auth
method and pending authentication via \fBauth_pending_file\fP\&.
.UNINDENT
.TP
.B  \-\-use\-prediction\-resistance
Enable prediction resistance on mbed TLS\(aqs RNG.
.sp
Enabling prediction resistance causes the RNG to reseed in each call for
random. Reseeding this often can quickly deplete the kernel entropy
pool.
.sp
If you need this option, please consider running a daemon that adds
entropy to the kernel pool.
.TP
.BI \-\-user \ user
Change the user ID of the OpenVPN process to \fBuser\fP after
initialization, dropping privileges in the process. This option is
useful to protect the system in the event that some hostile party was
able to gain control of an OpenVPN session. Though OpenVPN\(aqs security
features make this unlikely, it is provided as a second line of defense.
.sp
By setting \fBuser\fP to an unprivileged user dedicated to run openvpn,
the hostile party would be limited in what damage they could cause. Of
course once you take away privileges, you cannot return them to an
OpenVPN session. This means, for example, that if you want to reset an
OpenVPN daemon with a \fBSIGUSR1\fP signal (for example in response to
a DHCP reset), you should make use of one or more of the \fB\-\-persist\fP
options to ensure that OpenVPN doesn\(aqt need to execute any privileged
operations in order to restart (such as re\-reading key files or running
\fBifconfig\fP on the TUN device).
.sp
NOTE: Previous versions of openvpn used \fBnobody\fP as the example
unpriviledged user. It is not recommended to actually use that user
since it is usually used by other system services already. Always
create a dedicated user for openvpn.
.TP
.BI \-\-writepid \ file
Write OpenVPN\(aqs main process ID to \fBfile\fP\&.
.UNINDENT
.SS Log options
.INDENT 0.0
.TP
.BI \-\-echo \ parms
Echo \fBparms\fP to log output.
.sp
Designed to be used to send messages to a controlling application which
is receiving the OpenVPN log output.
.TP
.B  \-\-errors\-to\-stderr
Output errors to stderr instead of stdout unless log output is
redirected by one of the \fB\-\-log\fP options.
.TP
.BI \-\-log \ file
Output logging messages to \fBfile\fP, including output to stdout/stderr
which is generated by called scripts. If \fBfile\fP already exists it will
be truncated. This option takes effect immediately when it is parsed in
the command line and will supersede syslog output if \fB\-\-daemon\fP
is also specified. This option is persistent over the entire
course of an OpenVPN instantiation and will not be reset by
\fBSIGHUP\fP, \fBSIGUSR1\fP, or \fB\-\-ping\-restart\fP\&.
.sp
Note that on Windows, when OpenVPN is started as a service, logging
occurs by default without the need to specify this option.
.TP
.BI \-\-log\-append \ file
Append logging messages to \fBfile\fP\&.  If \fBfile\fP does not exist, it will
be created. This option behaves exactly like \fB\-\-log\fP except that it
appends to rather than truncating the log file.
.TP
.B  \-\-machine\-readable\-output
Always write timestamps and message flags to log messages, even when
they otherwise would not be prefixed. In particular, this applies to log
messages sent to stdout.
.TP
.BI \-\-mute \ n
Log at most \fBn\fP consecutive messages in the same category. This is
useful to limit repetitive logging of similar message types.
.TP
.B  \-\-mute\-replay\-warnings
Silence the output of replay warnings, which are a common false alarm on
WiFi networks. This option preserves the security of the replay
protection code without the verbosity associated with warnings about
duplicate packets.
.TP
.B  \-\-suppress\-timestamps
Avoid writing timestamps to log messages, even when they otherwise would
be prepended. In particular, this applies to log messages sent to
stdout.
.TP
.BI \-\-syslog \ progname
Direct log output to system logger, but do not become a daemon. See
\fB\-\-daemon\fP directive above for description of \fBprogname\fP parameter.
.TP
.BI \-\-verb \ n
Set output verbosity to \fBn\fP (default \fB1\fP). Each level shows all
info from the previous levels. Level \fB3\fP is recommended if you want
a good summary of what\(aqs happening without being swamped by output.
.INDENT 7.0
.TP
.B \fB0\fP
No output except fatal errors.
.TP
.B \fB1\fP to \fB4\fP
Normal usage range.
.TP
.B \fB5\fP
Outputs \fBR\fP and \fBW\fP characters to the console for
each packet read and write, uppercase is used for TCP/UDP
packets and lowercase is used for TUN/TAP packets.
.TP
.B \fB6\fP to \fB11\fP
Debug info range (see \fBerrlevel.h\fP in the source code for
additional information on debug levels).
.UNINDENT
.UNINDENT
.SS Protocol options
.sp
Options in this section affect features available in the OpenVPN wire
protocol.  Many of these options also define the encryption options
of the data channel in the OpenVPN wire protocol.  These options must be
configured in a compatible way between both the local and remote side.
.INDENT 0.0
.TP
.BI \-\-allow\-compression \ mode
As described in the \fB\-\-compress\fP option, compression is a potentially
dangerous option.  This option allows controlling the behaviour of
OpenVPN when compression is used and allowed.
.sp
The \fBmode\fP argument can be one of the following values:
.INDENT 7.0
.TP
.B \fBasym\fP
OpenVPN will only \fIdecompress incoming packets\fP but \fInot compress
outgoing packets\fP\&.  This also allows migrating to disable compression
when changing both server and client configurations to remove
compression at the same time is not a feasible option.
.TP
.B \fBno\fP  (default)
OpenVPN will refuse any compression.  If data\-channel offloading
is enabled, OpenVPN will additionally also refuse compression
framing (stub).
.TP
.B \fByes\fP
\fBDEPRECATED\fP This option is an alias for \fBasym\fP\&. Previously
it did enable compression for outgoing packets, but OpenVPN never
compresses packets on send now.
.UNINDENT
.TP
.BI \-\-auth \ alg
Authenticate data channel packets and (if enabled) \fBtls\-auth\fP control
channel packets with HMAC using message digest algorithm \fBalg\fP\&. (The
default is \fBSHA1\fP ). HMAC is a commonly used message authentication
algorithm (MAC) that uses a data string, a secure hash algorithm and a
key to produce a digital signature.
.sp
The OpenVPN data channel protocol uses encrypt\-then\-mac (i.e. first
encrypt a packet then HMAC the resulting ciphertext), which prevents
padding oracle attacks.
.sp
If an AEAD cipher mode (e.g. GCM) is chosen then the specified \fB\-\-auth\fP
algorithm is ignored for the data channel and the authentication method
of the AEAD cipher is used instead. Note that \fBalg\fP still specifies
the digest used for \fBtls\-auth\fP\&.
.sp
In static\-key encryption mode, the HMAC key is included in the key file
generated by \fB\-\-genkey\fP\&. In TLS mode, the HMAC key is dynamically
generated and shared between peers via the TLS control channel. If
OpenVPN receives a packet with a bad HMAC it will drop the packet. HMAC
usually adds 16 or 20 bytes per packet. Set \fBalg=none\fP to disable
authentication.
.sp
For more information on HMAC see
\fI\%https://tools.ietf.org/html/rfc2104\fP
.TP
.BI \-\-cipher \ alg
This option should not be used any longer in TLS mode and still
exists for two reasons:
.INDENT 7.0
.IP \(bu 2
compatibility with old configurations still carrying it
around;
.IP \(bu 2
allow users connecting to OpenVPN peers older than 2.6.0
to have \fB\-\-cipher\fP configured the same way as the remote
counterpart. This can avoid MTU/frame size warnings.
.UNINDENT
.sp
Before 2.4.0, this option was used to select the cipher to be
configured on the data channel, however, later versions usually
ignored this directive in favour of a negotiated cipher.
Starting with 2.6.0, this option is always ignored in TLS mode
when it comes to configuring the cipher.
.sp
If you wish to specify the cipher to use on the data channel,
please see \fB\-\-data\-ciphers\fP (for regular negotiation) and
\fB\-\-data\-ciphers\-fallback\fP (for a fallback option when the

release/src/router/rc/openvpn.c
/*
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston,
 * MA 02111-1307 USA
 *
 */

/* RC OpenVPN code for Asuswrt-Merlin. */

#include "rc.h"

#include <sys/types.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <string.h>

#include <openvpn_config.h>
#include <openvpn_control.h>

#if 0
void _ovpn_update_routing_rules(int add_cmd, const int unit);
#endif


int ovpn_up_main(int argc, char **argv) {
	int unit;
	int vpnc_idx;

	if(argc < 3)
		return -1;

	unit = atoi(argv[1]);

	if (!strcmp(argv[2], "server"))
		ovpn_server_up_handler(unit);
	else if (!strcmp(argv[2], "client")) {
		ovpn_client_up_handler(unit);

#ifdef RTCONFIG_MULTILAN_CFG
		vpnc_idx = get_vpnc_idx_by_proto_unit(VPN_PROTO_OVPN, unit);

		// Set vpnc%d_dns nvram
		vpnc_ovpn_set_dns(unit);
		// Create vpnc%d_resolv config file
		_gen_vpnc_resolv_conf(vpnc_idx);
#endif
		update_resolvconf();

#ifdef RTCONFIG_MULTILAN_CFG
		// Create iproute2 rules
		update_sdn_by_vpnc(vpnc_idx);
#endif
	} else
		return -1;

	return 0;
}

int ovpn_down_main(int argc, char **argv) {
	int unit, restart_dnsmasq = 0;
	int vpnc_idx;

	if(argc < 3)
		return -1;

	unit = atoi(argv[1]);
	if (!strcmp(argv[2], "server"))
		ovpn_server_down_handler(unit);
	else if (!strcmp(argv[2], "client")) {
#ifdef RTCONFIG_MULTILAN_CFG
		vpnc_idx = get_vpnc_idx_by_proto_unit(VPN_PROTO_OVPN, unit);
		update_sdn_by_vpnc(vpnc_idx);
		// Remove vpnc%d_dns nvram
		clean_vpnc_setting_value(vpnc_idx);
#endif

#if 0	// for dev_policy
		// clean routing rule
		clean_routing_rule_by_vpnc_idx(vpnc_idx);	// from vpnc_fusion.c
#endif

		ovpn_client_down_handler(unit);

		update_resolvconf();
	} else
		return -1;

	return 0;
}


#if 0	// Can handle SDN and dev_policy
void _ovpn_update_routing_rules(int add_cmd, const int unit) {
#ifdef RTCONFIG_VPN_FUSION_SUPPORT_INTERFACE
	char cmd_str[8], priority[8];
	int policy_cnt, i;
	VPNC_DEV_POLICY dev_policy[MAX_DEV_POLICY] = {{0}};
#endif
	int vpnc_idx;

	vpnc_idx = get_vpnc_idx_by_proto_unit(VPN_PROTO_OVPN, unit);

#ifdef RTCONFIG_MULTILAN_CFG
	update_sdn_by_vpnc(vpnc_idx);
#endif

#if defined(RTCONFIG_VPN_FUSION_SUPPORT_INTERFACE)
	policy_cnt = vpnc_get_dev_policy_list(dev_policy, MAX_DEV_POLICY, 0);
	snprintf(cmd_str, sizeof(cmd_str), "%s", (add_cmd == 0 ? "del" : "add");
	snprintf(id_str, sizeof(id_str), "%d", IP_ROUTE_TABLE_ID_VPNC_BASE + vpnc_idx);
	snprintf(priority, sizeof(priority), "%d", IP_RULE_PREF_VPNC_POLICY_IF + vpnc_idx * 3);

	for (i = 0; i < policy_cnt; ++i) {
		if(dev_policy[i].active && dev_policy[i].vpnc_idx == vpnc_idx && dev_policy[i].iif[0] != '\0') {
			if (add_cmd)
				eval("ip", "rule", "del", "priority", priority);

			eval("ip", "rule", (add_cmd ? "add" : "del"), "iif", dev_policy[i].iif, "table", id_str, "priority", priority);
		}
	}
#endif
}
#endif


int ovpn_route_up_main(int argc, char **argv) {
	ovpn_client_route_up_handler();
	return 0;
}

int ovpn_route_pre_down_main(int argc, char **argv) {
	ovpn_client_route_pre_down_handler();
	return 0;
}


/* Wrappers for library functions */
void inline start_ovpn_eas() {
	ovpn_process_eas(1);
}

void inline stop_ovpn_eas() {
	ovpn_process_eas(0);
}

void inline stop_ovpn_client(int unit) {
	ovpn_stop_client(unit);
#ifdef RTCONFIG_MULTILAN_CFG
	_vpnc_ipset_destroy( get_vpnc_idx_by_proto_unit(VPN_PROTO_OVPN, unit) );
#endif
}

void inline start_ovpn_client(int unit) {
#ifdef RTCONFIG_MULTILAN_CFG
	_vpnc_ipset_create( get_vpnc_idx_by_proto_unit(VPN_PROTO_OVPN, unit) );
#endif
	ovpn_start_client(unit);
}

void inline stop_ovpn_server(int unit) {
	ovpn_stop_server(unit);
}

void inline start_ovpn_server(int unit) {
	ovpn_start_server(unit);
}
